File: app/src/main/kotlin/com/openclaw/ghostcrab/data/impl/GatewayConnectionManagerImpl.kt
package com.openclaw.ghostcrab.data.impl

import com.openclaw.ghostcrab.data.api.OpenClawApiClient
import com.openclaw.ghostcrab.domain.exception.GatewayAuthException
import com.openclaw.ghostcrab.domain.exception.GatewayException
import com.openclaw.ghostcrab.domain.exception.GatewayUnreachableException
import com.openclaw.ghostcrab.domain.model.AuthRequirement
import com.openclaw.ghostcrab.domain.model.GatewayConnection
import com.openclaw.ghostcrab.domain.repository.GatewayConnectionManager
import com.openclaw.ghostcrab.domain.repository.SettingsRepository
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext

/**
 * Production implementation of [GatewayConnectionManager].
 *
 * Thread-safe via [mutex]. The held [OpenClawApiClient] is replaced on each [connect]
 * and released on [disconnect].
 *
 * @param clientFactory Creates [OpenClawApiClient] instances. Injectable for testing.
 */
class GatewayConnectionManagerImpl(
    private val clientFactory: OpenClawApiClientFactory = DefaultClientFactory,
    private val settingsRepository: SettingsRepository = AlwaysBlockCleartextSettingsRepository,
) : GatewayConnectionManager {

    private val _connectionState = MutableStateFlow<GatewayConnection>(GatewayConnection.Disconnected)
    override val connectionState: StateFlow<GatewayConnection> = _connectionState.asStateFlow()

    private var activeClient: OpenClawApiClient? = null
    private val mutex = Mutex()

    override suspend fun probeAuth(url: String): AuthRequirement = withContext(Dispatchers.IO) {
        val allowCleartext = settingsRepository.allowCleartextPublicIPs.firstOrNull() ?: false
        val probe = clientFactory.unauthenticated(url, allowCleartext)
        try {
            probe.health() // throws GatewayUnreachableException if down
            try {
                probe.status()
                AuthRequirement.None
            } catch (e: GatewayAuthException) {
                AuthRequirement.Token
            }
        } finally {
            probe.close()
        }
    }

    override suspend fun connect(url: String, token: String?) = withContext(Dispatchers.IO) {
        mutex.withLock {
            // Disconnect any existing session silently
            activeClient?.close()
            activeClient = null

            _connectionState.value = GatewayConnection.Connecting(url)

            // Held outside the try so every failure path can release the half-built client;
            // it only becomes activeClient once status() has succeeded.
            var client: OpenClawApiClient? = null
            try {
                val allowCleartext = settingsRepository.allowCleartextPublicIPs.firstOrNull() ?: false
                val authReq = probeAuth(url)
                client = if (token != null) {
                    clientFactory.authenticated(url, token, allowCleartext)
                } else {
                    clientFactory.unauthenticated(url, allowCleartext)
                }
                val statusResponse = client.status()
                val isHttps = url.startsWith("https://", ignoreCase = true)

                activeClient = client
                _connectionState.value = GatewayConnection.Connected(
                    url = url,
                    displayName = statusResponse.displayName,
                    version = statusResponse.version,
                    authRequirement = authReq,
                    isHttps = isHttps,
                    capabilities = statusResponse.capabilities,
                    hardwareInfo = statusResponse.hardware,
                    tokenOrNull = token,
                )
            } catch (e: GatewayException) {
                failConnect(url, e, client)
                throw e
            } catch (e: CancellationException) {
                // Caller went away mid-handshake: release the half-built client and return to
                // the terminal state instead of leaving the UI on Connecting forever.
                releaseClients(client)
                _connectionState.value = GatewayConnection.Disconnected
                throw e
            } catch (@Suppress("TooGenericExceptionCaught") e: Exception) {
                // Anything the client layer did not classify (e.g. a URL the engine rejects)
                // must still land in Error, never a stuck Connecting.
                val wrapped = GatewayUnreachableException(url, e)
                failConnect(url, wrapped, client)
                throw wrapped
            }
        }
    }

    /** Closes [pending] (a client whose handshake failed) and whatever is currently active. */
    private fun releaseClients(pending: OpenClawApiClient?) {
        pending?.close()
        activeClient?.close()
        activeClient = null
    }

    private fun failConnect(url: String, cause: GatewayException, pending: OpenClawApiClient?) {
        releaseClients(pending)
        _connectionState.value = GatewayConnection.Error(url, cause)
    }

    override suspend fun disconnect() {
        mutex.withLock {
            activeClient?.close()
            activeClient = null
            _connectionState.value = GatewayConnection.Disconnected
        }
    }

    /** Returns the active client, or null if not connected. Internal use for repositories. */
    fun requireClient(): OpenClawApiClient =
        activeClient ?: error("No active gateway connection")
}

// ── Fail-safe default settings (no cleartext to public IPs) ──────────────────

private object AlwaysBlockCleartextSettingsRepository : com.openclaw.ghostcrab.domain.repository.SettingsRepository {
    override val allowCleartextPublicIPs = kotlinx.coroutines.flow.flowOf(false)
    override suspend fun setAllowCleartextPublicIPs(enabled: Boolean) = Unit
}

// ── Factory interface for testability ────────────────────────────────────────

interface OpenClawApiClientFactory {
    fun unauthenticated(baseUrl: String, allowCleartextPublicIPs: Boolean = false): OpenClawApiClient
    fun authenticated(baseUrl: String, token: String, allowCleartextPublicIPs: Boolean = false): OpenClawApiClient
}

object DefaultClientFactory : OpenClawApiClientFactory {
    override fun unauthenticated(baseUrl: String, allowCleartextPublicIPs: Boolean) =
        OpenClawApiClient.unauthenticated(baseUrl, allowCleartextPublicIPs)
    override fun authenticated(baseUrl: String, token: String, allowCleartextPublicIPs: Boolean) =
        OpenClawApiClient.authenticated(baseUrl, token, allowCleartextPublicIPs)
}

File: ios/GhostCrab/Data/Impl/GatewayConnectionManagerImpl.swift
import Foundation

// MARK: - Client factory

/// Factory abstraction for ``OpenClawAPIClient`` creation.
///
/// Injectable into ``GatewayConnectionManagerImpl`` so tests can substitute a
/// mock client without touching the network. Mirrors Kotlin's
/// `OpenClawApiClientFactory` interface.
public protocol OpenClawAPIClientFactory: Sendable {
    func unauthenticated(baseURL: URL, allowCleartextPublicIPs: Bool) -> OpenClawAPIClient
    func authenticated(baseURL: URL, token: String, allowCleartextPublicIPs: Bool) -> OpenClawAPIClient
}

/// Default factory used in production — forwards to the static factories on
/// ``OpenClawAPIClient`` itself.
public struct DefaultOpenClawAPIClientFactory: OpenClawAPIClientFactory {
    public init() {}

    public func unauthenticated(baseURL: URL, allowCleartextPublicIPs: Bool) -> OpenClawAPIClient {
        OpenClawAPIClient.unauthenticated(baseURL: baseURL, allowCleartextPublicIPs: allowCleartextPublicIPs)
    }

    public func authenticated(baseURL: URL, token: String, allowCleartextPublicIPs: Bool) -> OpenClawAPIClient {
        OpenClawAPIClient.authenticated(baseURL: baseURL, token: token, allowCleartextPublicIPs: allowCleartextPublicIPs)
    }
}

// MARK: - Connection manager

/// Production implementation of ``GatewayConnectionManager``.
///
/// `actor` provides the same mutual exclusion as the Kotlin `Mutex` — every
/// `connect` / `disconnect` / `probeAuth` call is serialized through the actor.
/// The held ``OpenClawAPIClient`` is replaced on each ``connect(url:token:)`` and
/// released on ``disconnect()``.
///
/// The connection state is exposed via ``connectionStates()`` as an
/// `AsyncStream<GatewayConnection>`. Each call returns an independent stream
/// backed by the same underlying state — the latest value is replayed on
/// subscribe (mirroring Kotlin's `StateFlow.value`), and every subsequent
/// transition is yielded to all live subscribers. A synchronous snapshot is
/// available via ``currentState``.
///
/// Direct port of `GatewayConnectionManagerImpl.kt`.
public actor GatewayConnectionManagerImpl: GatewayConnectionManager {

    // ── State ────────────────────────────────────────────────────────────────

    /// Latest connection state. Read by ``currentState`` and replayed to every
    /// new subscriber.
    private var state: GatewayConnection = .disconnected

    /// Active subscribers — fanned out on every state transition. Removed on
    /// continuation termination.
    private var subscribers: [UUID: AsyncStream<GatewayConnection>.Continuation] = [:]

    /// Currently held client, or `nil` when disconnected. Replaced/released by
    /// ``connect(url:token:)`` and ``disconnect()``.
    private var activeClient: OpenClawAPIClient?

    private let clientFactory: any OpenClawAPIClientFactory
    private let settingsRepository: any SettingsRepository

    // ── Init ─────────────────────────────────────────────────────────────────

    /// Designated initializer.
    ///
    /// - Parameters:
    ///   - clientFactory: Creates ``OpenClawAPIClient`` instances. Injectable
    ///     for testing; defaults to ``DefaultOpenClawAPIClientFactory``.
    ///   - settingsRepository: Source for the `allowCleartextPublicIPs`
    ///     preference. Defaults to a stub that always blocks cleartext —
    ///     fail-safe identical to the Kotlin
    ///     `AlwaysBlockCleartextSettingsRepository`.
    public init(
        clientFactory: any OpenClawAPIClientFactory = DefaultOpenClawAPIClientFactory(),
        settingsRepository: any SettingsRepository = AlwaysBlockCleartextSettingsRepository()
    ) {
        self.clientFactory = clientFactory
        self.settingsRepository = settingsRepository
    }

    // ── GatewayConnectionManager: state ──────────────────────────────────────

    /// Synchronous snapshot of the current connection state.
    ///
    /// Lock-protected via ``SnapshotBox`` — does not hop through the actor, so
    /// callers can read it without `await`. Equivalent to Kotlin's
    /// `connectionState.value`. Writes happen only inside ``publish(_:)``
    /// (actor-isolated), so update ordering is preserved.
    public nonisolated var currentState: GatewayConnection {
        // The protocol declares this as a synchronous computed property, so we
        // must service the read without a hop. Each transition publishes the
        // value into an actor-isolated property *and* a nonisolated
        // synchronization primitive — see ``snapshotBox`` below.
        snapshotBox.value
    }

    /// Returns an `AsyncStream` that replays the latest state on subscribe and
    /// emits every subsequent transition. Each call yields an independent
    /// stream — multiple consumers are supported.
    public nonisolated func connectionStates() -> AsyncStream<GatewayConnection> {
        AsyncStream { continuation in
            let id = UUID()

            // Register first, *then* yield the latest snapshot from inside the
            // actor. This guarantees the subscriber's first event is the state
            // observed at registration time — no missed-transition race
            // between subscribe and the next publish.
            Task { [weak self] in
                guard let self else {
                    continuation.finish()
                    return
                }
                await self.registerAndReplay(id: id, continuation: continuation)
            }

            continuation.onTermination = { [weak self] _ in
                guard let self else { return }
                Task { await self.unregister(id: id) }
            }
        }
    }

    // ── GatewayConnectionManager: actions ────────────────────────────────────

    /// Probes auth requirements without storing any state.
    ///
    /// `/health` failure → ``GatewayError/unreachable``.
    /// `/status` 401/403 → `.token`. Otherwise → `.none`.
    ///
    /// - Parameter url: Base URL of the gateway.
    /// - Returns: The inferred ``AuthRequirement``.
    /// - Throws: ``GatewayError`` propagated from the probe.
    public func probeAuth(url: String) async throws -> AuthRequirement {
        let baseURL = try parseURL(url)
        let allowCleartext = await firstAllowCleartext()
        let probe = clientFactory.unauthenticated(
            baseURL: baseURL,
            allowCleartextPublicIPs: allowCleartext
        )
        do {
            _ = try await probe.health()
            do {
                _ = try await probe.status()
                await probe.close()
                return .none
            } catch let e as GatewayError {
                if case .auth = e {
                    await probe.close()
                    return .token
                }
                throw e
            }
        } catch {
            await probe.close()
            throw error
        }
    }

    /// Connects to a gateway. See ``GatewayConnectionManager/connect(url:token:)``.
    public func connect(url: String, token: String?) async throws {
        // Silently drop any prior session — same semantics as Kotlin.
        if let prior = activeClient {
            await prior.close()
            activeClient = nil
        }

        let baseURL = try parseURL(url)
        publish(.connecting(url: url))

        do {
            let allowCleartext = await firstAllowCleartext()
            let authReq = try await probeAuth(url: url)
            let client: OpenClawAPIClient
            if let token {
                client = clientFactory.authenticated(
                    baseURL: baseURL,
                    token: token,
                    allowCleartextPublicIPs: allowCleartext
                )
            } else {
                client = clientFactory.unauthenticated(
                    baseURL: baseURL,
                    allowCleartextPublicIPs: allowCleartext
                )
            }
            let statusResponse = try await client.status()
            let isHttps = url.lowercased().hasPrefix("https://")

            activeClient = client
            publish(.connected(
                url: url,
                displayName: statusResponse.displayName,
                version: statusResponse.version,
                authRequirement: authReq,
                isHttps: isHttps,
                capabilities: statusResponse.capabilities,
                hardwareInfo: statusResponse.hardware,
                tokenOrNull: token
            ))
        } catch let e as GatewayError {
            if let pending = activeClient {
                await pending.close()
                activeClient = nil
            }
            publish(.error(url: url, cause: e))
            throw e
        } catch {
            // Anything not already a GatewayError gets normalized to
            // .unreachable so callers see a typed error — same posture as
            // Kotlin's `safeRequest` catch-all.
            let mapped = GatewayError.unreachable(url: url, underlying: error as? (any Error & Sendable))
            if let pending = activeClient {
                await pending.close()
                activeClient = nil
            }
            publish(.error(url: url, cause: mapped))
            throw mapped
        }
    }

    /// Disconnects from the current gateway. No-op if already disconnected.
    public func disconnect() async {
        if let prior = activeClient {
            await prior.close()
            activeClient = nil
        }
        publish(.disconnected)
    }

    // ── Internal: client access for repository impls ─────────────────────────

    /// Returns the active client, or throws if disconnected.
    ///
    /// Equivalent to Kotlin's `requireClient()` — repository impls call this
    /// to obtain the live ``OpenClawAPIClient``. Throws a typed error rather
    /// than a Kotlin `IllegalStateException`.
    ///
    /// - Returns: The currently active ``OpenClawAPIClient``.
    /// - Throws: ``GatewayError/unreachable`` with a synthetic URL if no
    ///   connection is active. The choice of `.unreachable` matches the
    ///   "no active session" mental model: from the caller's perspective the
    ///   gateway is, for all intents, unreachable.
    public func requireClient() throws -> OpenClawAPIClient {
        guard let client = activeClient else {
            throw GatewayError.unreachable(url: "<no active connection>", underlying: nil)
        }
        return client
    }

    // ── Subscriber management ────────────────────────────────────────────────

    /// Registers a subscriber and yields the current actor-isolated state in
    /// the same atomic step — guarantees no transitions are missed between
    /// subscribe and the first emission.
    private func registerAndReplay(
        id: UUID,
        continuation: AsyncStream<GatewayConnection>.Continuation
    ) {
        subscribers[id] = continuation
        continuation.yield(state)
    }

    private func unregister(id: UUID) {
        subscribers.removeValue(forKey: id)
    }

    /// Publishes a new state to every subscriber and the snapshot box.
    ///
    /// The snapshot box is written first so a synchronous `currentState` read
    /// racing with the state transition observes the new value before any
    /// subscriber.
    private func publish(_ next: GatewayConnection) {
        state = next
        snapshotBox.set(next)
        for cont in subscribers.values {
            cont.yield(next)
        }
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    private func firstAllowCleartext() async -> Bool {
        for await value in settingsRepository.allowCleartextPublicIPs() {
            return value
        }
        return false
    }

    private nonisolated func parseURL(_ raw: String) throws -> URL {
        guard let url = URL(string: raw) else {
            throw GatewayError.api(url: raw, statusCode: 0, body: "Malformed URL")
        }
        return url
    }

    // MARK: - Snapshot box

    /// Synchronous mailbox for the current state.
    ///
    /// Kotlin exposes `StateFlow<GatewayConnection>` whose `.value` is a
    /// non-suspending read. The Swift protocol mirrors that by declaring
    /// `currentState` as a synchronous computed property — which the actor
    /// cannot satisfy without a `nonisolated` escape hatch. ``SnapshotBox``
    /// is a `Sendable` lock-protected cell whose updates are sequenced by
    /// the actor (writes only happen inside ``publish(_:)``) but whose reads
    /// are synchronous and lock-only.
    private let snapshotBox = SnapshotBox()
}

// ── Snapshot box impl ────────────────────────────────────────────────────────

/// Lock-protected cell holding the latest ``GatewayConnection`` for synchronous
/// snapshot reads from any thread.
///
/// Necessary because ``GatewayConnectionManager/currentState`` is declared as a
/// synchronous computed property on the protocol — the actor cannot service
/// non-isolated reads of its mutable state without this indirection.
private final class SnapshotBox: @unchecked Sendable {
    private var stored: GatewayConnection = .disconnected
    private let lock = NSLock()

    var value: GatewayConnection {
        lock.lock(); defer { lock.unlock() }
        return stored
    }

    func set(_ next: GatewayConnection) {
        lock.lock(); defer { lock.unlock() }
        stored = next
    }
}

// ── Fail-safe default settings (no cleartext to public IPs) ──────────────────

/// Stub ``SettingsRepository`` that always reports cleartext-public-IPs as
/// disabled. Used as a fail-safe default in
/// ``GatewayConnectionManagerImpl/init(clientFactory:settingsRepository:)``.
///
/// Direct port of Kotlin's `AlwaysBlockCleartextSettingsRepository`.
public struct AlwaysBlockCleartextSettingsRepository: SettingsRepository {
    public init() {}

    public func allowCleartextPublicIPs() -> AsyncStream<Bool> {
        AsyncStream { continuation in
            continuation.yield(false)
            continuation.finish()
        }
    }

    public func setAllowCleartextPublicIPs(_ enabled: Bool) async {
        // no-op
    }
}
